internal/services/pam/pam_test.go

```diff
@@ -0,0 +1,606 @@
+	optionalWithBooleans := "optional:true,false"
+
+	normalLayout := &authd.UILayout{
+		Type:    "form",
+		Label:   &required,
+		Button:  &optional,
+		Wait:    &optionalWithBooleans,
+		Entry:   &supportedEntries,
+		Content: &optional,
+	}
+	invalidLayout := &authd.UILayout{
+		Type: "",
+	}
+
+	tests := map[string]struct {
+		brokerID  string
+		sessionID string
+		username  string
+		layout    *authd.UILayout
+
+		wantErr bool
+	}{
+		"Successfully get authentication modes": {},
+
+		"Error when sessionID is empty":           {sessionID: "empty", wantErr: true},
+		"Error when passing invalid layout":       {sessionID: "invalid layout", layout: invalidLayout, wantErr: true},
+		"Error when broker does not exist":        {sessionID: "no broker", brokerID: "does not exist", wantErr: true},
+		"Error when getting authentication modes": {username: "GAM_error", wantErr: true},
+		"Error when broker returns invalid modes": {username: "GAM_invalid", wantErr: true},
+	}
+	for name, tc := range tests {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			brokerCfgDir := t.TempDir()
+			brokerCfg := testutils.StartBusBrokerMock(t, brokerCfgDir)
+
+			client := startClient(t, brokerCfgDir)
+
+			if tc.brokerID == "" {
+				tc.brokerID = getGeneratedID(t, client, filepath.Base(brokerCfg))
+			}
+
+			if tc.brokerID != "does not exist" {
+				if tc.username == "" {
+					tc.username = "does not matter"
+				}
+				sbResp, err := client.SelectBroker(context.Background(), &authd.SBRequest{
+					BrokerId: tc.brokerID,
+					Username: tc.username,
+				})
+				require.NoError(t, err, "Setup: failed to create session for tests")
+				if tc.sessionID == "" {
+					tc.sessionID = sbResp.GetSessionId()
+				}
+			}
+
+			if tc.layout == nil {
+				tc.layout = normalLayout
+			}
+
+			if tc.sessionID == "empty" {
+				tc.sessionID = ""
+			}
+
+			gamReq := &authd.GAMRequest{
+				SessionId:          tc.sessionID,
+				SupportedUiLayouts: []*authd.UILayout{tc.layout},
+			}
+			gamResp, err := client.GetAuthenticationModes(context.Background(), gamReq)
+			if tc.wantErr {
+				require.Error(t, err, "GetAuthenticationModes should return an error, but did not")
+				return
+			}
+			require.NoError(t, err, "GetAuthenticationModes should not return an error, but did")
+			got := gamResp.GetAuthenticationModes()
+			want := testutils.LoadWithUpdateFromGoldenYAML(t, got)
+			require.Equal(t, want, got, "GetAuthenticationModes returned an unexpected response")
+		})
+	}
+}
+
+func TestSelectAuthenticationMode(t *testing.T) {
+	t.Parallel()
+
+	tests := map[string]struct {
+		username  string
+		sessionID string
+		brokerID  string
+		authMode  string
+
+		wantErr bool
+	}{
+		"Successfully select form":        {username: "SAM_form_success"},
+		"Successfully select qrcode":      {username: "SAM_qrcode_success"},
+		"Successfully select newpassword": {username: "SAM_newpassword_success"},
+
+		// service errors
```

Comment:
There is one missing (main to me) use case: Failing when the selected auth mode does not exist.

Reply:
yeah, something like selecting unexisting?